### PR TITLE
Use std variant

### PR DESCRIPTION
**Pull request by joakim-hove**
All the SVD routines can specify the number of singular values to retain in two different ways - either as an integer specifying the number of singular values directly, or as a floating point number specifying the amount variation explained.

With this PR this is implemented with `std::variant<double, int>`

On top of #2681

### CODE REVIEW

**Comment by joakim-hove** [comment]
Test ert please!

File: libres/lib/analysis/enkf_linalg.cpp [hunk 1]
```diff
@@ -93,60 +93,55 @@ void enkf_linalg_genX2(matrix_type *X2, const matrix_type *S,
 /*This function is similar to enkf_linalg_svdS but it returns the eigen values without its inverse and also give the matrices truncated U VT and Sig0*/
 
 // Trunc.SVD(S)  = U0 * Sig0 * V0T
-int enkf_linalg_svd_truncation(const matrix_type *S, double truncation,
-                               int ncomp, dgesvd_vector_enum store_V0T,
-                               double *sig0, matrix_type *U0,
-                               matrix_type *V0T) {
+int enkf_linalg_svd_truncation(const matrix_type *S,
```
Comment (dafeda) [inline libres/lib/analysis/enkf_linalg.cpp:96]: We removed this function in the following commit:

https://github.com/equinor/ert/commit/ba43cfb51327e4a227b580c0cfb9753e7ad63662

File: libres/lib/analysis/enkf_linalg.cpp [hunk 2]
```diff
@@ -93,60 +93,55 @@ void enkf_linalg_genX2(matrix_type *X2, const matrix_type *S,
...
+                               const std::variant<double, int> &truncation,
+                               dgesvd_vector_enum store_V0T, double *sig0,
+                               matrix_type *U0, matrix_type *V0T) {
 
     int num_significant = -1;
     int nrows = matrix_get_rows(S);
     int ncolumns = matrix_get_columns(S);
 
-    if (((truncation > 0) && (ncomp < 0)) ||
```
Comment (dafeda) [inline libres/lib/analysis/enkf_linalg.cpp:105]: Are we comfortable removing this check?

**Comment by dafeda** [comment]
What is the main reason of replacing two input parameters with one `std::variant<double, int>`?

File: libres/lib/analysis/enkf_linalg.cpp (hunk 1 above)
Comment (joakim-hove) [reply libres/lib/analysis/enkf_linalg.cpp:96]: I'll rebase

File: libres/lib/analysis/enkf_linalg.cpp (hunk 2 above)
Comment (joakim-hove) [reply libres/lib/analysis/enkf_linalg.cpp:105]: Yes - the point about using `std::variant<>` is to enforce this from the compiler

Comment (dafeda) [reply libres/lib/analysis/enkf_linalg.cpp]: I see, bit I don't think there is anything stopping us from passing in a negative truncation.
Perhaps this is a non-issue?

**Comment by dafeda** [comment]
> > What is the main reason of replacing two input parameters with one std::variant<double, int>?
> 
> You can specify the number of singular values to retain in two _mutually exclusive_ ways:
> 
> 1. Using a floating point truncation - and include singular values up to that fraction of variation explained.
> 2. Specify an integer number of singular values.
> 
> Before this PR this was handled by using
> 
> ```
> double truncation;
> int num_singular;
> ```
> 
> and then a system of using invalid (i.e. `-1`) values for the inactive alternative. When using `std::variant<double,int>`the compiler enforces for us that exactly one of the options is active.

That makes good sense, thanks.

That description would also make a great commit message.

File: libres/lib/analysis/enkf_linalg.cpp (hunk 2 above)
Comment (joakim-hove) [reply libres/lib/analysis/enkf_linalg.cpp]: > I see, bit I don't think there is anything stopping us from passing in a negative truncation.

No - in principle I agree with that, but that should caught when the truncation is set.

Comment (dafeda) [reply libres/lib/analysis/enkf_linalg.cpp]: I agree.

File: libres/lib/include/ert/analysis/enkf_linalg.hpp [hunk 3]
```diff
@@ -22,7 +19,13 @@ void enkf_linalg_init_sqrtX(matrix_type *X5, const matrix_type *S,
 void enkf_linalg_Cee(matrix_type *B, int nrens, const matrix_type *R,
                      const matrix_type *U0, const double *inv_sig0);
 
-int enkf_linalg_svdS(const matrix_type *S, double truncation, int ncomp,
+int enkf_linalg_svd_truncation(const matrix_type *S,
```
Comment (dafeda) [inline libres/lib/include/ert/analysis/enkf_linalg.hpp:22]: Thing this can be removed.

**Review by dafeda — APPROVED** [review]
Please remember to squash commits, and do add a description in the commit message (in addition to the usual one-liner).

LGTM 👍

File: libres/lib/include/ert/analysis/enkf_linalg.hpp (hunk 3 above)
Comment (joakim-hove) [reply libres/lib/include/ert/analysis/enkf_linalg.hpp]: OK

**Comment by joakim-hove** [comment]
test ert please

**Pull request merged**